src/async-wrap.cc

```diff
@@ -282,37 +285,87 @@ bool AsyncWrap::EmitAfter(Environment* env, double async_id) {
 class PromiseWrap : public AsyncWrap {
  public:
   PromiseWrap(Environment* env, Local<Object> object, bool silent)
-    : AsyncWrap(env, object, PROVIDER_PROMISE, silent) {}
+    : AsyncWrap(env, object, PROVIDER_PROMISE, silent) {
+    MakeWeak(this);
+  }
   size_t self_size() const override { return sizeof(*this); }
+
+  static constexpr int kPromiseField = 1;
+  static constexpr int kParentIdField = 2;
+  static constexpr int kInternalFieldCount = 3;
+
+  static PromiseWrap* New(Environment* env,
+                          Local<Promise> promise,
+                          PromiseWrap* parent_wrap,
+                          bool silent);
+  static void GetPromise(Local<String> property,
+                         const PropertyCallbackInfo<Value>& info);
+  static void GetParentId(Local<String> property,
+                          const PropertyCallbackInfo<Value>& info);
 };
 
+PromiseWrap* PromiseWrap::New(Environment* env,
+                              Local<Promise> promise,
+                              PromiseWrap* parent_wrap,
+                              bool silent) {
+  Local<Object> object = env->promise_wrap_template()
+                            ->NewInstance(env->context()).ToLocalChecked();
+  object->SetInternalField(PromiseWrap::kPromiseField, promise);
+  if (parent_wrap != nullptr) {
+    object->SetInternalField(PromiseWrap::kParentIdField,
+                             Number::New(env->isolate(),
+                                         parent_wrap->get_id()));
+  }
+  CHECK_EQ(promise->GetAlignedPointerFromInternalField(0), nullptr);
+  promise->SetInternalField(0, object);
+  return new PromiseWrap(env, object, silent);
+}
+
+void PromiseWrap::GetPromise(Local<String> property,
+                             const PropertyCallbackInfo<Value>& info) {
+  info.GetReturnValue().Set(info.Holder()->GetInternalField(kPromiseField));
+}
+
+void PromiseWrap::GetParentId(Local<String> property,
+                              const PropertyCallbackInfo<Value>& info) {
+  info.GetReturnValue().Set(info.Holder()->GetInternalField(kParentIdField));
+}
 
 static void PromiseHook(PromiseHookType type, Local<Promise> promise,
                         Local<Value> parent, void* arg) {
   Local<Context> context = promise->CreationContext();
   Environment* env = Environment::GetCurrent(context);
-  PromiseWrap* wrap = Unwrap<PromiseWrap>(promise);
+  Local<Value> resource_object_value = promise->GetInternalField(0);
+  Local<Object> resource_object;
+  PromiseWrap* wrap = nullptr;
+  if (resource_object_value->IsObject()) {
+    resource_object = resource_object_value.As<Object>();
```

Comment:
In addition a Debug build it asserts, so instead of simply crashing b/c it was cast to the wrong type it'll actually say you've screwed up.